File: Add Code Here/C++/Rat-in-Maze.cpp
class Solution{
    private:
    
    bool issafe(int x,int y,int n,vector<vector<int>>visited,vector<vector<int>> &m){
        if((x>=0&&x<n)&&(y>=0&&y<n)&&visited[x][y]==0&&m[x][y]==1) return true;
        return false;
    }
    
    void solve(vector<vector<int>> &m, int n,vector<vector<int>>visited, vector<string>&ans,int x,int y,string s){
        if(x==n-1&&y==n-1){
            ans.push_back(s);
            return;
        }
        visited[x][y]=1;
        
        // down
        int newx=x+1;
        int newy=y;
        if(issafe(newx,newy,n,visited,m)){
            s.push_back('D');
            solve(m,n,visited,ans,newx,newy,s);
            s.pop_back();
        }
        // left
        newx=x;
        newy=y-1;
        if(issafe(newx,newy,n,visited,m)){
            s.push_back('L');
            solve(m,n,visited,ans,newx,newy,s);
            s.pop_back();
        }
        // right
        newx=x;
        newy=y+1;
        if(issafe(newx,newy,n,visited,m)){
            s.push_back('R');
            solve(m,n,visited,ans,newx,newy,s);
            s.pop_back();
        }
        // Up
        newx=x-1;
        newy=y;
        if(issafe(newx,newy,n,visited,m)){
            s.push_back('U');
            solve(m,n,visited,ans,newx,newy,s);
            s.pop_back();
        }
        visited[x][y]=0;
    }
    
    public:
    vector<string> findPath(vector<vector<int>> &m, int n) {
        vector<vector<int>> visited=m;
        string s="";
        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                visited[i][j]=0;
            }
        }
        vector<string>ans;
        if(m[0][0]==0) return ans;
        int srcx=0;
        int srcy=0;
        solve(m,n,visited,ans,srcx,srcy,s);
        sort(ans.begin(),ans.end());
        return ans;
    }
};

File: Add Code Here/Java/Rat_In_Maze.java
public class RatInMaze {

    private static final int OPEN = 1;
    private static final int PATH = 2;
    private int[][] maze;
    private int[][] solution;

    public RatInMaze(int[][] maze) {
        this.maze = maze;
        this.solution = new int[maze.length][maze[0].length];
    }

    public boolean solve() {
        if (findPath(0, 0)) {
            printSolution();
            return true;
        } else {
            System.out.println("No solution exists.");
            return false;
        }
    }

    private boolean findPath(int x, int y) {
        // Check if we reached the destination
        if (x == maze.length - 1 && y == maze[0].length - 1) {
            solution[x][y] = PATH;
            return true;
        }

        if (isSafe(x, y)) {
            solution[x][y] = PATH;

            // Move forward in X direction
            if (findPath(x + 1, y)) {
                return true;
            }

            // Move down in Y direction
            if (findPath(x, y + 1)) {
                return true;
            }

            // Backtrack if no direction leads to a solution
            solution[x][y] = 0;
        }

        return false;
    }

    private boolean isSafe(int x, int y) {
        return x >= 0 && x < maze.length && y >= 0 && y < maze[0].length && maze[x][y] == OPEN;
    }

    private void printSolution() {
        for (int i = 0; i < solution.length; i++) {
            for (int j = 0; j < solution[i].length; j++) {
                System.out.print(solution[i][j] + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        int[][] maze = {
            {1, 0, 0, 0},
            {1, 1, 0, 1},
            {0, 1, 0, 0},
            {1, 1, 1, 1}
        };
        RatInMaze rat = new RatInMaze(maze);
        rat.solve();
    }
}
